### PR TITLE
sql: enable indexing and ordering on arrays of orderable and indexable types

### LINKED ISSUE #17154: sql: allow ARRAYs in forward indexes
Arrays support inverted indexes, but not forward indexes. If you want to index arrays, starting in CockroachDB 20.1 consider using an inverted index on your column.

This issue tracks adding forward indexes to array types.

Jira issue: CRDB-6049

### LINKED ISSUE #35707: sql: support ordering by array columns
Currently it's not possible to order a table by an array column.

This is because arrays do not have a valid lexicographic ordering.

for reference, PostgreSQL supports it.

### PR DESCRIPTION
**Pull request by rohany**
Fixes #17154.
Fixes #35707.

This PR enables arrays to be ordered and indexed by
introducing an ordered key encoding for arrays.
Once this exists, the rest of the SQL infrastructure
is ready to handle indexing and ordering on arrays.

To encode an array of elements `ARRAY[a, b]`,
we create the following encoding.

Let `AM` = a marker byte for arrays and let `AT` be a terminator byte.

`enc(ARRAY[a, b]) = [AM, enc(a), enc(b), AT]`

The key is that the terminator is less than the element marker.
This allows for the "prefix matching" style comparison that
arrays support.

Release note (sql change): This PR adds support for indexing
and ordering of arrays of indexable and orderable inner types.

### CODE REVIEW

**Review by rytaft — COMMENTED** [review]
> Reviewed 2 of 11 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, and @rohany)

___
*[pkg/sql/opt/optbuilder/testdata/orderby, line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5r72OU3Q3bENcmGI8N:-M5r72OU3Q3bENcmGI8O:bnyt8by) ([raw file](https://github.com/cockroachdb/cockroach/blob/7811ade70322d2f6a1b5888686f76f36aecd6086/pkg/sql/opt/optbuilder/testdata/orderby#L391)):*
> ```
> error (42P01): no data source matches prefix: a
> 
> build
> ```

Can you also add a test here for arrays of type json?

**Comment by rohany** [comment]
We don't currently support arrays of json

**Review by jordanlewis — COMMENTED** [review]
Nice, this will be a great improvement!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-:-M5rs0j-BRT1fRz2dNu2:bleu0yg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, and @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/array, line 1352 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5rqkgiCf3a3CD4V51v:-M5rqkgiCf3a3CD4V51w:bgoiqi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7811ade70322d2f6a1b5888686f76f36aecd6086/pkg/sql/logictest/testdata/logic_test/array#L1352)):*
> ```
> # TODO (rohany): We have always done this for array comparisons, so I think
> #  it would be a breaking change + opposite with our other null behavior to
> #  change it suddenly...
> ```

I agree. Can you check that Postgres changes the null sorting behavior within arrays when you change NULLS FIRST / NULLS LAST?
___
*[pkg/sql/sqlbase/column_type_encoding.go, line 927 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5rrTG623uLUfPrMY-X:-M5rrTG623uLUfPrMY-Y:b-ygfl7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/7811ade70322d2f6a1b5888686f76f36aecd6086/pkg/sql/sqlbase/column_type_encoding.go#L927)):*
> ```Go
> // encodeArrayKey generates an ordered key encoding of an array.
> // The encoding format for an array [a, b] is as follows:
> // [arrayMarker, elementMarker, enc(a), elementMarker, enc(b), terminator]
> ```

I don't understand why we need the markers. Wouldn't the sorting work just as well if we simply did `[enc(a), enc(b)]` in your example?

It looks like arrays are supposed to sort lexicographically - it's element by element comparison, and longer arrays lose. But I think that that would already be the case with normal key encoding:

`[enc(a), enc(b)]` would be less than `[enc(a), enc(b), anything else]` just by the normal lexicographical sorting that keys use.

Does NULL somehow break this?

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, and @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/array, line 1352 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5rqkgiCf3a3CD4V51v:-M5rsYfXBEzwND0xDH8Y:bgrkeqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7811ade70322d2f6a1b5888686f76f36aecd6086/pkg/sql/logictest/testdata/logic_test/array#L1352)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I agree. Can you check that Postgres changes the null sorting behavior within arrays when you change NULLS FIRST / NULLS LAST?
</blockquote></details>

It doesn't seem to:
```
rohany=# create table t (x int[]);
CREATE TABLE
rohany=# insert into t values (array[1]), (array[null::int]);
INSERT 0 2
rohany=# select * from t order by x
rohany-# ;
   x
--------
 {1}
 {NULL}
(2 rows)

rohany=# select * from t order by x NULLS first;
   x
--------
 {1}
 {NULL}
(2 rows)

rohany=# select * from t order by x NULLS last;
   x
--------
 {1}
 {NULL}
(2 rows)
```
___
*[pkg/sql/sqlbase/column_type_encoding.go, line 927 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5rrTG623uLUfPrMY-X:-M5rsqeA8Kpajqg2ofLe:b-7slpgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7811ade70322d2f6a1b5888686f76f36aecd6086/pkg/sql/sqlbase/column_type_encoding.go#L927)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I don't understand why we need the markers. Wouldn't the sorting work just as well if we simply did `[enc(a), enc(b)]` in your example?

It looks like arrays are supposed to sort lexicographically - it's element by element comparison, and longer arrays lose. But I think that that would already be the case with normal key encoding:

`[enc(a), enc(b)]` would be less than `[enc(a), enc(b), anything else]` just by the normal lexicographical sorting that keys use.

Does NULL somehow break this?
</blockquote></details>

Yeah, the issue is NULLs. Nulls are encoded as the byte `0x00`. If we don't have the markers, then null would conflict with the array terminator. But we can't just add a marker for NULL only, because then it wouldn't sort right against non-null elements. Am I missing something straightforward here?

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/array, line 1352 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5rqkgiCf3a3CD4V51v:-M5vsxOn-vVrWp2mOvLN:b-oid2lg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7811ade70322d2f6a1b5888686f76f36aecd6086/pkg/sql/logictest/testdata/logic_test/array#L1352)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

It doesn't seem to:
```
rohany=# create table t (x int[]);
CREATE TABLE
rohany=# insert into t values (array[1]), (array[null::int]);
INSERT 0 2
rohany=# select * from t order by x
rohany-# ;
   x
--------
 {1}
 {NULL}
(2 rows)

rohany=# select * from t order by x NULLS first;
   x
--------
 {1}
 {NULL}
(2 rows)

rohany=# select * from t order by x NULLS last;
   x
--------
 {1}
 {NULL}
(2 rows)
```
</blockquote></details>

Ok. That's a little bit unfortunate as it might mean that we're supposed to do this no matter what. But I guess I agree that we cant change our behavior now anyway.
___
*[pkg/sql/sqlbase/column_type_encoding.go, line 927 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5rrTG623uLUfPrMY-X:-M5vtJu88INdmY1m2ADy:b4oepw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7811ade70322d2f6a1b5888686f76f36aecd6086/pkg/sql/sqlbase/column_type_encoding.go#L927)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Yeah, the issue is NULLs. Nulls are encoded as the byte `0x00`. If we don't have the markers, then null would conflict with the array terminator. But we can't just add a marker for NULL only, because then it wouldn't sort right against non-null elements. Am I missing something straightforward here?
</blockquote></details>

I'm not sure yet.  By the way how would this interact with multi-column keys? Maybe my lexicographic ordering comment was wrong, since of course keys can be longer than just a single column. How does the ordering distinguish between short and long columns when there could be a second column afterward?

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @otan)

___
*[pkg/sql/logictest/testdata/logic_test/array, line 1352 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5rqkgiCf3a3CD4V51v:-M5w2ej56iQF0jYnFmzf:bwlw3c5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7811ade70322d2f6a1b5888686f76f36aecd6086/pkg/sql/logictest/testdata/logic_test/array#L1352)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Ok. That's a little bit unfortunate as it might mean that we're supposed to do this no matter what. But I guess I agree that we cant change our behavior now anyway.
</blockquote></details>

We'd have to do something a little wonky like switch the null encoding direction when encoding arrays to support this
___
*[pkg/sql/sqlbase/column_type_encoding.go, line 927 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5rrTG623uLUfPrMY-X:-M5w7Kl5CGAlfzqB70N-:br96e5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/7811ade70322d2f6a1b5888686f76f36aecd6086/pkg/sql/sqlbase/column_type_encoding.go#L927)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I'm not sure yet.  By the way how would this interact with multi-column keys? Maybe my lexicographic ordering comment was wrong, since of course keys can be longer than just a single column. How does the ordering distinguish between short and long columns when there could be a second column afterward?
</blockquote></details>

there is an array terminator byte, so thats how we know. I'll add some tests for it though.

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rohany)

___
*[pkg/sql/sqlbase/column_type_encoding.go, line 927 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5rrTG623uLUfPrMY-X:-M5wD6um30RxxfUn67oq:b-q2l6si) ([raw file](https://github.com/cockroachdb/cockroach/blob/7811ade70322d2f6a1b5888686f76f36aecd6086/pkg/sql/sqlbase/column_type_encoding.go#L927)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

there is an array terminator byte, so thats how we know. I'll add some tests for it though.
</blockquote></details>

I see.

Could we encode null as a byte that's greater than `0x00`, but less than all other values? It looks like we could use `0x01`, which is reserved for `encodedNotNull` which is never actually used in a key.

**Comment by rohany** [comment]
PTAL, updated the encoding strategy to not use the extra marker bytes between elements, and instead use a different null encoding within the array

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/array, line 1319 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5yvOjx-cOmMZ3XOMc1:-M5yvOjx-cOmMZ3XOMc2:b78buhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ee173932d3c3f1164a7607f5b36cc50f027328d/pkg/sql/logictest/testdata/logic_test/array#L1319)):*
> ```
> 
> subtest array_indexes
> ```

Other testing ideas:

arrays w/ composite types (decimal)
arrays w/ strings? cuz their encoding is escape based - if we changed our decoder to try to search for an endpoint it would fail
more than 1 array in a single index
group by, distinct, join on array (or hmm we already added those, right?)
foreign key to array
lookup join to array? i'm not sure if these are necessary... just thinking

can you add execbuilder tests for the array scans that show the spans that get generated for array range queries?
___
*[pkg/util/encoding/encoding.go, line 1294 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5ytIkSCccWzWo-lrRO:-M5ytIkSCccWzWo-lrRP:b-gu4x2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ee173932d3c3f1164a7607f5b36cc50f027328d/pkg/util/encoding/encoding.go#L1294)):*
> ```Go
> 	TimeTZ       Type = 19
> 	Geo          Type = 20
> 	ArrayKeyAsc  Type = 21 // Array key encoding
> ```

But what about the existing `Array` byte? I still don't understand why we can't reuse it.

Also, how come most of these other fields don't have `Desc` equivalents? What makes array/string/etc special?

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @otan)

___
*[pkg/sql/logictest/testdata/logic_test/array, line 1319 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5yvOjx-cOmMZ3XOMc1:-M6149zh7hG4aUhhwcwR:bsudmrn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ee173932d3c3f1164a7607f5b36cc50f027328d/pkg/sql/logictest/testdata/logic_test/array#L1319)):*
> arrays w/ composite types (decimal)

already included

> arrays w/ strings? cuz their encoding is escape based - if we changed our decoder to try to search for an endpoint it would fail

Done, added bytes too.

> more than 1 array in a single index

Done

> group by, distinct, join on array (or hmm we already added those, right?)

groupby and distinct are tested, added join tests

> foreign key to array

Done -- this helped to catch a bug!

> lookup join to array? i'm not sure if these are necessary... just thinking

Done

> can you add execbuilder tests for the array scans that show the spans that get generated for array range queries?

done
___
*[pkg/sql/sqlbase/column_type_encoding.go, line 927 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5rrTG623uLUfPrMY-X:-M61jJ9QDNpyLbyym8BF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7811ade70322d2f6a1b5888686f76f36aecd6086/pkg/sql/sqlbase/column_type_encoding.go#L927)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I see.

Could we encode null as a byte that's greater than `0x00`, but less than all other values? It looks like we could use `0x01`, which is reserved for `encodedNotNull` which is never actually used in a key.
</blockquote></details>

Done.
___
*[pkg/util/encoding/encoding.go, line 1294 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M5ytIkSCccWzWo-lrRO:-M61jJbsCq_1uJOPntZP:b7lofpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ee173932d3c3f1164a7607f5b36cc50f027328d/pkg/util/encoding/encoding.go#L1294)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

But what about the existing `Array` byte? I still don't understand why we can't reuse it.

Also, how come most of these other fields don't have `Desc` equivalents? What makes array/string/etc special?
</blockquote></details>

RE the existing Array byte -- I wanted to get rid of this, but there are some places where we lose information when we do that. For example, when attempting to pretty print an array key, we need to know whether it is a key or value encoding of an array, where if we just use the existing Array type we won't have that information.

RE the others, I think the decoding logic can just figure out what to do based on the encodings themselves.  If anything, i think that these type bytes are mostly exported views of the internal marker bytes.

**Comment by awoods187** [comment]
I'm excited about this work!

**Review by jordanlewis — COMMENTED** [review]
Ok looks good but let's just get rid of the desc tag and type everywhere - we can do as the other types do when they need descending encoding.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-:-M6bW-zB3_HAQ7CSHHCN:b2pb21b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, and @rohany)

___
*[pkg/sql/rowexec/version_history.txt, line 114 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M6aliXN0Fb-_Fk_aZEV:-M6aliXN0Fb-_Fk_aZEW:bbyyqto) ([raw file](https://github.com/cockroachdb/cockroach/blob/7206021a3cb9a6a8a2d616d6a71b8118e692c526/pkg/sql/rowexec/version_history.txt#L114)):*
> ```
> - Version: 29 (MinAcceptedVersion: 29)
>     - The hashing of DArray datums was changed and is incompatible with the
>       existing method of hasing DArrays.
> ```

nit: s/hasing/hashing/
___
*[pkg/sql/sqlbase/column_type_encoding_test.go, line 213 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M6amWBD69ZVUXQVp1eY:-M6amWBD69ZVUXQVp1eZ:bdo03yo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7206021a3cb9a6a8a2d616d6a71b8118e692c526/pkg/sql/sqlbase/column_type_encoding_test.go#L213)):*
> ```Go
> 	// Also run the property on arrays possibly containing NULL values.
> 	// The random generator in the property above does not generate NULLs.
> 	properties.Property("order-preserving-arrays", prop.ForAll(
> ```

Heh how do you like using this Gopter thing
___
*[pkg/util/encoding/encoding.go, line 88 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M6bVP8O2xUMh80s9PVX:-M6bVP8O2xUMh80s9PVY:b-k4yetb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7206021a3cb9a6a8a2d616d6a71b8118e692c526/pkg/util/encoding/encoding.go#L88)):*
> ```Go
> 	// Markers and terminators for key encoding Datum arrays in sorted order.
> 	arrayKeyMarker                    = geoMarker + 1
> 	arrayKeyDescendingMarker          = arrayKeyMarker + 1
> ```

I think that we don't need these descending markers. See discussion here: https://cockroachlabs.slack.com/archives/C0HM2DZ34/p1588718172047500?thread_ts=1588717530.044900&cid=C0HM2DZ34

The tl;dr is that the descending thing isn't necessary since we no longer have a self-describing key format. The fact that bytesDesc exists was an early attempt and isn't strictly necessary.

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @otan)

___
*[pkg/sql/sqlbase/column_type_encoding_test.go, line 213 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M6amWBD69ZVUXQVp1eY:-M6bX8Qd8qOJbXdAj_U5:bww4g9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/7206021a3cb9a6a8a2d616d6a71b8118e692c526/pkg/sql/sqlbase/column_type_encoding_test.go#L213)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Heh how do you like using this Gopter thing
</blockquote></details>

lol its pretty nice -- tho it's definitely super wonky to start with.
___
*[pkg/util/encoding/encoding.go, line 88 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M6bVP8O2xUMh80s9PVX:-M6baUQk8A1MZx83CBR9:b-jj3j2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/7206021a3cb9a6a8a2d616d6a71b8118e692c526/pkg/util/encoding/encoding.go#L88)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think that we don't need these descending markers. See discussion here: https://cockroachlabs.slack.com/archives/C0HM2DZ34/p1588718172047500?thread_ts=1588717530.044900&cid=C0HM2DZ34

The tl;dr is that the descending thing isn't necessary since we no longer have a self-describing key format. The fact that bytesDesc exists was an early attempt and isn't strictly necessary.
</blockquote></details>

added a comment about this

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rohany)

___
*[pkg/sql/sqlbase/column_type_encoding.go, line 927 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/48045#-M6kZMzo2gg2_uvXtLd0:-M6kZMzo2gg2_uvXtLd1:b-yxd6id) ([raw file](https://github.com/cockroachdb/cockroach/blob/60cb535037c74e964abb0696b4ec7e68861181b0/pkg/sql/sqlbase/column_type_encoding.go#L927)):*
> ```Go
> // encodeArrayKey generates an ordered key encoding of an array.
> // The encoding format for an array [a, b] is as follows:
> // [arrayMarker, enc(a), enc(b), terminator].
> ```

I'm afraid this is still going to cause problems because of string encoding.

In fact, I'm confused how it didn't already break your string test cases. Encoded strings are terminated with the sequence \x00\x01. But our terminator is also \x00, isn't it? What am I missing...

**Comment by rohany** [comment]
The important thing is the context in what is attempting to be decoded. When we are decoding a string, we know we are looking for \x00\x01 as the terminator. once we return from decoding a string, we go back to the context of an array, and interpret \x00 as a terminator. Additionally, the string encodings escape \x00 themselves.

**Comment by jordanlewis** [comment]
Oh yeah. True! And our Skip/GetLength functions actually iterate over the whole thing - not searching for the end directly. I do feel like that's not ideal... it would be better to have a fixed end so skipping / finding length could be just a simple bytes search. But maybe it's fine, since we don't even expect that we'll be seeing so many array keys.

**Comment by rohany** [comment]
Well they can't skip, because we can't encode the length of the bytes and strings at the front of the string/byte sequence. It seems they use some faster assembly implemented function to do it (bytes.IndexByte or something).

**Comment by jordanlewis** [comment]
Right, I'm saying we could use `IndexByte` if we went with a scheme where there was a fixed end byte. But I guess this is probably not worth it.

**Comment by rohany** [comment]
Ohh, I see what you mean. It would be tricky if that byte lived within other encoded data though -- we'd need our own escaping as well.

**Comment by jordanlewis** [comment]
Right. I guess it's not worth it.

LGTM :shipit:
